### ISSUE TITLE
Regression in Plusminus training

### ISSUE THREAD

**Issue by Shreeshrii** [tesseract-ocr/tesseract#1798]
From Wiki
```
Ray - 100 iterations -   1.26% char 3.98% word error 
Ray - 3600 iterations - 0.041%       0.185% 
Ray - Training Eval - 0.0326%       0.128%
Ray - Impact Eval - 2.3767074       8.3829474
Ray - Grep - ...you should see that it gets all the ± signs correct! 
```
----------------------------------------------------------------------------------
4.00alpha - 2018-01-05 (000d027) Egor Pugin: Rename tesseract library.
```
alpha 100- char train=0.55%, word train=1.856%, 
alpha 3600 - char train=0.031%, word train=0.069%
alpha - training eval - Eval Char error rate=0.010627193, Word error rate=0.028920449
alpha - Impact Eval - Eval Char error rate=3.2334834, Word error rate=8.8460551
```

4.0.0-beta.1 -  40f4311 - Beta release for Ubuntu 18.04
```
beta.1 100-  char train=0.5%, word train=1.693%, 
beta.1 3600 - char train=0.029%, word train=0.1%,
beta.1 - training eval -  Eval Char error rate=0.012183638, Word error rate=0.039235013
beta.1 - Impact Eval - Eval Char error rate=3.2191941, Word error rate=8.6803766
```

 4.0.0-beta.2 -  10f4998 - Ubuntu 18.04 long term release
```
beta.2 100- char train=0.534%, word train=1.787%,
beta.2 3600 - char train=0.028%, word train=0.072%
beta.2 - training eval -  Eval Char error rate=0.019994024, Word error rate=0.069248312
beta.2 - Impact Eval -  Eval Char error rate=2.8908476, Word error rate=7.8440695
```

4.0.0-beta.3 -  b1f7990 - 4.0.0-beta.3 Release
```
beta.3 100-  char train=1.16%, word train=3.331%,
beta.3 3600 - char train=0.143%, word train=0.389%,
beta.3 - training eval - Eval Char error rate=0.086619098, Word error rate=0.21720323
beta.3 - Impact Eval - Eval Char error rate=5.3591565, Word error rate=12.885069
```

4.0.0 beta - current code
```
current - 100
current - 3600 - char train=0.107%, word train=0.297%,
```

**Comment by Shreeshrii** [comment]
Attached are the commands used for the plusminus training tutorial
[plusminus-tutorial.sh.txt](https://github.com/tesseract-ocr/tesseract/files/2220473/plusminus-tutorial.sh.txt)
.

**Comment by Shreeshrii** [comment]
Change seems to be between beta.2 and beta.3. Error rate has doubled for for the first 100 iterations of training itself.

**Comment by stweil** [comment]
I try to reproduce the results and still cannot see in which step of the script I should see the numbers and output shown above.

**Comment by stweil** [comment]
One of the steps is preparing shuffled `eng.plusminusnew.training_text` text file. The order of text lines then is randomly, so some variations in output when running the same command sequence can be expected.

**Comment by Shreeshrii** [comment]
@stweil The numbers are copied from the display at end of processing, copied for sake of easy comparison.

In my repo, I have added the 14 lines with plusminus sign and shuffled randomly and saved under the new name. For the batch for testing that I posted. I added the shuf command so that all lines with plusminus sign are not all together at end.

**Comment by Shreeshrii** [comment]
Results for 100 iterations and 3600 iterations will be at end of following command. You can change debug level to 0 for fewer displays on console.

```
#----------------------------
time lstmtraining \
  --debug_interval -1 \
  --model_output ../tesstutorial/trainplusminus/plusminus \
  --continue_from ../tesstutorial/trainplusminus/eng.lstm \
  --traineddata ../tesstutorial/trainplusminus/eng/eng.traineddata \
  --old_traineddata ../tessdata_best/eng.traineddata \
  --train_listfile ../tesstutorial/trainplusminus/eng.training_files.txt \
  --max_iterations 100
#----------------------------
time lstmtraining \
  --debug_interval -1 \
  --model_output ../tesstutorial/trainplusminus/plusminus \
  --continue_from ../tesstutorial/trainplusminus/eng.lstm \
  --traineddata ../tesstutorial/trainplusminus/eng/eng.traineddata \
  --old_traineddata ../tessdata_best/eng.traineddata \
  --train_listfile ../tesstutorial/trainplusminus/eng.training_files.txt \
  --max_iterations 3600
```

**Comment by Shreeshrii** [comment]
lstmeval results are shown  after msg:

```
Loaded 99/99 pages (1-99) of document ../tesstutorial/trainplusminus/eng.Verdana_Italic.exp0.lstmf
Truth:FREE was <a top days Login this 2004 & - said first 27 then 2. $100 they FIG. [1]
OCR  :FREE was <a top days Login this 2004 8 - said first 27 then 2. $100 they FIG. [1]
At iteration 0, stage 0, Eval Char error rate=0.018135086, Word error rate=0.057594972

real    21m3.924s
user    69m45.320s
sys     0m14.700s
```

Eval Char error rate=0.018135086, Word error rate=0.057594972

**Comment by stweil** [comment]
So my result is even worse:

    At iteration 401/3600/3600, Mean rms=0.173%, delta=0.051%, char train=0.167%, word train=0.481%, skip ratio=0%,  New worst char error = 0.167 wrote checkpoint.

**Comment by stweil** [comment]
@Shreeshrii, do you also get different results when running the same training steps several times? For the results shown above I always used the same shuffled training text. In addition, I disabled multi-threading in some runs. Even then, the results vary, so there must be some other factors which cause the somewhat random results. That's bad for comparing results and would also be bad to use it in continuous integration tests.

**Comment by Shreeshrii** [comment]
I did not run these training steps multiple time for same commit and note
results. However a slight variation is quite possible because the
lines/pages from lstmf files are used for training in random order. (There
is a variable for sequential training recommended to be used if training
for just one kind of images, as per training wiki.)

I would still expect the results to be within a certain range.

In my single runs of beta.2 and beta.3, there seems to be larger
difference. There have also been user reports that not all cases of
plusminus are recognized after training.

Do you also see a difference between these commits or eg. Beta2 and current
code?

For continuous integration I would suggest an error/warning if results are
more than 5-10% different from those of beta.2 (Ubuntu release). You can
decide on the appropriate % to be used based on the variance.

As an alternative or as a crosscheck, you can use tesseract to recognize
the image using the new traineddata and check the accuracy using an
independent OCR evaluation tool.

**Comment by Shreeshrii** [comment]
Randomized Training Data and sequential_training
For Stochastic Gradient Descent to work properly, the training data is supposed to be randomly shuffled across all the sample files, so the trainer can read its way through each file in turn and go back to the first one when it reaches the end. This is entirely contrary to the way base Tesseract was trained!

If using the rendering code, (via tesstrain.sh) then it will shuffle the sample text lines within each file, but you will get a set of files, each containing training samples from a single font. To add a more even mix, the default is to process one sample from each file in turn aka 'round robin' style. If you have generated training data some other way, or it is all from the same style (a handwritten manuscript book for instance) then you can use the --sequential_training flag for lstmtraining. This is more memory efficient since it will load data from only two files at a time, and process them in sequence. (The second file is read-ahead so it is ready when needed.)

**Comment by stweil** [comment]
> Change seems to be between beta.2 and beta.3.

Still good: tesseract 4.0.0-beta.2-153-g7dea.

**Comment by Shreeshrii** [comment]
@stweil Thanks for testing to confirm whether there is regression.

Please also see discussion regarding finetuning at https://groups.google.com/forum/m/?utm_medium=email&utm_source=footer#!msg/tesseract-ocr/COJ4IjcrL6s/GnvIpZ2uBgAJ

Is it possible that the plus minus best model is occurring before 3600 iterations?

**Comment by stweil** [comment]
According to my test, commit fdc243b3638fa799cae3b921e2b137d6121425b1 caused the regression. I will run a last test to confirm this with git master and a revert of that commit.

**Comment by amitdo** [comment]
>    common_args+=" --leading=${LEADING} --xsize 2550"

try to change 2550 to 2560.

**Comment by stweil** [comment]
Commit a27e91c4f945e3b0962a8465c452938f3320596e already did that. The regression remains.

**Comment by Shreeshrii** [comment]
I had changed the size  to avoid errors relating to image too large to
train.

https://github.com/tesseract-ocr/tesseract/blob/fbff323d6a260bf80ae8e84da9df47cf0dbd491f/src/lstm/lstmrecognizer.cpp#L225

**Comment by amitdo** [comment]
I don't think there is a real issue here.

`n_iter` number of iterations of `n_lines` lines in length `line_len` should not be compared to `n_iter` number of iterations of `n_lines` lines in length `line_len_shorter`.

The difference is expected. If the lines are shorter, then more iteration will be needed.

**Comment by stweil** [comment]
I just looked at the images. The text part seems to be identical – they only differ in their width, but that is simply white background. Why does more or less white background make a difference for training?

**Comment by stweil** [comment]
Update: I was partially wrong. The text part is identical if the text fits into the (shorter) line. Long text lines with wide letters don't fit and are wrapped. Then the rendered lines no longer match the lines of the original text. This could be really bad for the training process if that tries to match line images to wrong text lines. If I'm right, that would also explain the bad results with reduced image width and why the results get better when the width is increased.

**Comment by amitdo** [comment]
What about doing the opposite?

**Comment by stweil** [comment]
What do you mean by "opposite"? @Shreeshrii reduced the image width to 2560 which caused the "regression", maybe because some text lines don't fit into that width.

**Comment by Shreeshrii** [comment]
>This could be really bad for the training process if that tries to match
line images to wrong text lines

You can take a look at the box files to see whether there is a tab to mark
EOL at end of shorter line.

**Comment by Shreeshrii** [comment]
>What about increasing kMaxImageWidth from 2560 to 4096 (OpenCL seems to want a multiple of 512), for example?

A larger image width may have other effects on LSTM training that we don't know about.

**Comment by Shreeshrii** [comment]
You can revert fdc243b to fix the regression.

**Comment by Shreeshrii** [comment]
@stweil Thank you for the investigation and fix for the issue.
